### DISCUSSION TITLE
zustand and fetching

### DISCUSSION THREAD

**Discussion by Sebastian-Nielsen** [pmndrs/zustand#2538] [Q&A]
How would you go about having a zustand store that stores the result of a fetch when you are under these circumstances:

1. The fetch that populates the zustand store is not triggered on app load. 
2. The zustand store should store the fetched result in a format that complies to `Async<T>`. 

(Ideally, the zustand store should use the immer middleware)
 

```ts
export type Async<T> = {
	loadingStatus: 'success',
	data: T,
	error: undefined,
} | {
	loadingStatus: 'loading',
	data: undefined,
	error: undefined,
} | {
	loadingStatus: 'error'
	data: undefined,
	error: any,
};
```

At this point, I think it's near impossible to have my zustand store conform to the `Async<T>` type. I am hoping someone can prove me wrong or at the very least guide me in the right direction. 😀

**Comment by dai-shi** [comment]
It might not be as beautiful as you'd expect:
```js
const useFooStore = create((set) => ({
  result: { loadingStatus: 'loading' } as Async<MyType>,
  doFetch: () => /* ... */,
}));
```
